runtime/dynstats.c

```diff
@@ -499,12 +601,151 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
 	RETiRet;
 }
 
+/* not supported in older platforms - disable */
+#ifndef O_CLOEXEC
+#define O_CLOEXEC 0
+#endif
+/*
+ *	Helper function adapted from imfile of the same name - possible candidate for refactor.
+ */
+static rsRetVal ATTR_NONNULL()
+atomicWriteStateFile(const char *fn, const char *content) {
+	DEFiRet;
+	const int fd = open(fn, O_CLOEXEC | O_NOCTTY | O_WRONLY | O_CREAT | O_TRUNC, 0600);
+	if(fd < 0) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: cannot open state file '%s' for "
+			"persisting file state - some data will probably be duplicated "
+			"on next startup", fn);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+	const size_t toWrite = strlen(content);
+	const ssize_t w = write(fd, content, toWrite);
+	if(w != (ssize_t) toWrite) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: partial write to state file '%s' "
+			"this may cause trouble in the future. We will try to delete the "
+			"state file, as this provides most consistent state", fn);
+		unlink(fn);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+finalize_it:
+	if(fd >= 0) {
+		close(fd);
+	}
+	RETiRet;
+}
+
+static const uchar *
+getStateFileDir(dynstats_bucket_t *b)
+{
+	const uchar *wrkdir;
+	assert(b != NULL);
+	if (b->stateFileDirectory == NULL) {
+		wrkdir = glblGetWorkDirRaw();
```

Comment:
Should we use a subdir?

Reply:
We do already have the ability for user specify a working directory already, so do we still need it?

Reply:
I think it can make sense to permit different locations other than the global work dir - sometimes it makes sense to split different state info for easier management. But if I understood @janmejay correctly, he is suggesting to think about a subdir under the work dir.

Reply:
Ok, perhaps we can add a subdir like 'dynstats' underneath the global work dir.